src/kakarot/instructions/system_operations.cairo

```diff
@@ -39,63 +39,124 @@ namespace SystemOperations {
     }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
         alloc_locals;
 
+        let memory = ctx.memory;
+        let state = ctx.state;
+        let stack = ctx.stack;
+
+        let opcode_number = [ctx.call_context.bytecode + ctx.program_counter];
+        let is_create2 = is_not_zero(opcode_number - 0xf0);
+        let popped_len = 3 + is_create2;
+        let (stack, popped) = Stack.pop_n(stack, 3 + is_create2);
+
+        let value = popped[0];
+        let offset = popped[1];
+        let size = popped[2];
+
+        // Charge dynamic gas
+        let memory_expansion_cost = Memory.expansion_cost(memory, offset.low + size.low);
+        // If .high != 0, OOG is surely triggered. So we only use the .low part for the
+        // actually computation, and add ctx.call_context.gas_limit * .high which would
+        // either be 0 or ctx.call_context.gas_limit * k, thus triggering OOG.
+        let memory_expansion_cost = ctx.call_context.gas_limit * (offset.high + size.high) +
+            memory_expansion_cost;
+        let (calldata_words, _) = unsigned_div_rem(size.low + 31, 31);
+        let init_code_gas = 2 * calldata_words;
+        let calldata_word_gas = is_create2 * 6 * calldata_words;
+        let ctx = ExecutionContext.charge_gas(
+            ctx, memory_expansion_cost + init_code_gas + calldata_word_gas
+        );
+
+        if (ctx.reverted != FALSE) {
+            let ctx = ExecutionContext.update_stack(ctx, stack);
+            return ctx;
+        }
+
+        // Get message call gas
+        let available_gas = ctx.call_context.gas_limit - ctx.gas_used;
+        let (gas_limit, _) = unsigned_div_rem(available_gas, 64);
```

Comment:
why is this called gas limit ? it's 1/63th of the available gas right?

Reply:
it's just that I cannot put in the same line `available_gas - unsigned_div_rem(available_gas, 64).quotient` so I've found this convenient to just put the target name directly though it's still partial computation